contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,861 @@
+
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+import { UniswapV2Library } from "../../external/contracts/UniswapV2Library.sol";
+import { SushiswapV2Library } from "../../external/contracts/SushiswapV2Library.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IBasicIssuanceModule } from "../interfaces/IBasicIssuanceModule.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+
+/**
+ * @title ExchangeIssuance
+ * @author Index Coop
+ *
+ * Contract for minting and redeeming any Set token using
+ * ETH or an ERC20 as the paying/receiving currency. All swaps are done using the best price
+ * found on Uniswap or Sushiswap.
+ *
+ */
+contract ExchangeIssuance is ReentrancyGuard {
+    
+    // TODO: use safeERC20
+    // TODO: Discuss KNC token approve function.
+    
+    using SafeMath for uint256;
+    
+    /* ============ Enums ============ */
+    
+    enum Exchange { Uniswap, Sushiswap }
+    
+    /* ============ constants ============ */
+    
+    uint256 constant private MAX_UINT96 = 2 ** 96 - 1;
+    
+    /* ============ State Variables ============ */
+
+    IUniswapV2Router02 private uniRouter;
+    address private uniFactory;
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _inputToken,
+        uint256 _amountIn,
+        uint256 _amountOut
+    );
+    event ExchangeRedeem(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _outputToken,
+        uint256 _amountIn,
+        uint256 _amountOut
+    );
+
+    /* ============ Constructor ============ */
+
+    constructor(
+        address _uniFactory,
+        IUniswapV2Router02 _uniRouter, 
+        address _sushiFactory, 
+        IUniswapV2Router02 _sushiRouter, 
+        IBasicIssuanceModule _basicIssuanceModule
+    )
+        public
+    {
+        uniFactory = _uniFactory;
+        uniRouter = _uniRouter;
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = _sushiRouter;
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = _basicIssuanceModule;
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+    
+    /* ============ Public Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a set token during a 
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token) public {
```

Comment:
Why is this public and not external?

Reply:
It is being used by other two approve functions.